### PR TITLE
Adding max shortcut

### PR DESCRIPTION
**Pull request by Megamiun**
Related to #162

----
I confirm that I have read the [Contributor Agreements v1.0](https://github.com/robstoll/atrium/blob/master/.github/Contributor%20Agreements%20v1.0.txt), agree to be bound on them and confirm that my contribution is compliant.

### CODE REVIEW

File: apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/iterableAssertions.kt [hunk 1]
```diff
@@ -151,6 +151,24 @@ fun <E : Any, T : Iterable<E?>> Expect<T>.containsExactly(
...
+    ExpectImpl.iterable.max(this).getExpectOfFeature()
+
+/**
+ * Expects that the subject of the assertion (an [Iterable]) contains an entry holding
+ * the assertions created by [assertionCreator]
+ *
+ * @return This assertion container to support a fluent API.
+ * @throws AssertionError Might throw an [AssertionError] if the assertion made is not correct.
+ */
+fun <E : Comparable<E>,  T : Iterable<E?>> Expect<T>.max(assertionCreator: Expect<E>.() -> Unit): Expect<T> =
```
Comment (robstoll) [inline apis/fluent-en_GB/atrium-api-fluent-en_GB-common/src/main/kotlin/ch/tutteli/atrium/api/fluent/en_GB/iterableAssertions.kt]: `E?` is wrong IMO

File: domain/robstoll-lib/atrium-domain-robstoll-lib-common/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/iterableAssertions.kt [hunk 2]
```diff
@@ -32,6 +33,17 @@ fun <E, T : Iterable<E>> _containsBuilder(subjectProvider: SubjectProvider<T>):
 fun <E, T : Iterable<E>> _containsNotBuilder(subjectProvider: SubjectProvider<T>): IterableContains.Builder<E, T, NotSearchBehaviour> =
     IterableContainsBuilder(subjectProvider, NotSearchBehaviourImpl())
 
+fun <E : Comparable<E>, T : Iterable<E?>> _max(assertionContainer: Expect<T>): ExtractedFeaturePostStep<T, E> {
+    return ExpectImpl.feature.extractor(assertionContainer)
+        .withDescription("max")
+        .withRepresentationForFailure(DescriptionIterableAssertion.EMPTY_ITERABLE)
+        .withCheck { it.filterNotNull().isNotEmpty() }
+        .withFeatureExtraction {
+            it.filterNotNull().max()
```
Comment (robstoll) [inline domain/robstoll-lib/atrium-domain-robstoll-lib-common/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/iterableAssertions.kt]: Same here don't filter null. Null is only returned if the iterable was empty

File: misc/specs/atrium-specs-common/src/main/kotlin/ch/tutteli/atrium/specs/integration/IterableComparableAssertionsSpec.kt [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        it("fails warning about empty iterable") {
+            expect {
+                emptyIterable.testMax { toBe(1) }
+            }.toThrow<AssertionError> {
+                messageContains("cannot be determined, empty Iterable")
+            }
+        }
+    }
+
+    context("list of nulls") {
```
Comment (robstoll) [inline misc/specs/atrium-specs-common/src/main/kotlin/ch/tutteli/atrium/specs/integration/IterableComparableAssertionsSpec.kt]: Null is a good use case but actually not possible as Kotlin's max only accepts `T : Comparable<T>` and comparable is invariant (as far as I know, check if you think I am wrong)

File: translations/en_GB/atrium-translations-en_GB-common/src/main/kotlin/ch/tutteli/atrium/translations/DescriptionIterableAssertion.kt [hunk 4]
```diff
@@ -14,6 +14,7 @@ enum class DescriptionIterableAssertion(override val value: String) : StringBase
     AT_MOST("is at most"),
     CONTAINS("contains"),
     CONTAINS_NOT("does not contain"),
+    EMPTY_ITERABLE("!! cannot be determined, empty Iterable"),
```
Comment (robstoll) [inline translations/en_GB/atrium-translations-en_GB-common/src/main/kotlin/ch/tutteli/atrium/translations/DescriptionIterableAssertion.kt]: There should be NO_ELEMENTS by now which provides the de_CH translation in addition. Reuse it

**Comment by Megamiun** [comment]
Ok, I think I made all the changes

File: domain/robstoll-lib/atrium-domain-robstoll-lib-common/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/iterableAssertions.kt [hunk 5]
```diff
@@ -80,14 +80,24 @@ fun <E : Any> _hasNotNext(expect: Expect<Iterable<E>>): Assertion =
...
 fun <E : Comparable<E>, T : Iterable<E>> _min(assertionContainer: Expect<T>): ExtractedFeaturePostStep<T, E> =
-    ExpectImpl.feature.extractor(assertionContainer)
-        .methodCall("min")
-        .withRepresentationForFailure(DescriptionIterableAssertion.NO_ELEMENTS)
+    collect(assertionContainer, "min", Iterable<E>::min)
+
+fun <E : Comparable<E>, T : Iterable<E>> _max(assertionContainer: Expect<T>): ExtractedFeaturePostStep<T, E> =
+    collect(assertionContainer, "max", Iterable<E>::max)
+
+private fun <E : Comparable<E>, T : Iterable<E>> collect(
```
Comment (robstoll) [inline domain/robstoll-lib/atrium-domain-robstoll-lib-common/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/iterableAssertions.kt]: 🙂👍

File: misc/specs/atrium-specs-common/src/main/kotlin/ch/tutteli/atrium/specs/integration/IterableFeatureAssertionsSpec.kt [hunk 6]
```diff
@@ -74,4 +92,44 @@ abstract class IterableFeatureAssertionsSpec(
...
+            }
+        }
+    }
+
+    context("empty list") {
+        it("fails warning about empty iterable") {
+            expect {
+                emptyIterable.testMax { toBe(1) }
+            }.toThrow<AssertionError> {
+                messageContains("cannot be determined, empty Iterable")
```
Comment (robstoll) [inline misc/specs/atrium-specs-common/src/main/kotlin/ch/tutteli/atrium/specs/integration/IterableFeatureAssertionsSpec.kt]: Please use the translatable instead (see line 73)

**Comment by robstoll** [comment]
One last small thing and I can merge

**Comment by robstoll** [comment]
@Megamiun thanks for the contribution :+1:
You have another issue assigned. Let me know in case you don't want to do it. One to go for a hat-trick.

**Comment by robstoll** [comment]
@Megamiun would be nice if you could take a look why the coverage dropped by 10% I thought its an codecov error on the PR jut it is now the same on master.

**Comment by Megamiun** [comment]
About the other assigned issue: I think I will drop it, is seems that to use a Spek2, we would have to remake all tests that need to be forgiven as Spek2 instead of Spek.

About the coverage, I will try to see about it, but it seems very strange

**Comment by robstoll** [comment]
Coverage is fixed again, no need to investigate. Please un-assign yourself if you don't plan to continue. ReadmeTestEngine forgives as well, you can have a look how it is done there

**Pull request merged**